<div class="clearfix" style="padding: 10px; padding-left: 0px">
<a href="http://bombora.com"><img src="https://app.box.com/shared/static/e0j9v1xjmubit0inthhgv3llwnoansjp.png" width="200px" class="pull-right" style="display: inline-block; margin: 5px; vertical-align: middle;"></a>
<h1> Bombora Data Science: <br> *Interview Exam* </h1>
</div>

<img width="200px" src="https://app.box.com/shared/static/15slg1mvjd1zldbg3xkj9picjkmhzpa5.png">

---
# Welcome

Welcome! This notebook contains interview exam questions referenced in the *Instructions* section in the `README.md`—please read that first, *before* attempting to answer questions here.

<div class="alert alert-info" role="alert" style="margin: 10px">
<p style="font-weight:bold">ADVICE</p>
<p>*Do not* read these questions, and panic, *before* reading the instructions in `README.md`.</p>
</div>

<div class="alert alert-warning" role="alert" style="margin: 10px">
<p style="font-weight:bold">WARNING</p>

<p>If using <a href="https://try.jupyter.org">try.jupyter.org</a> do not rely on the server for anything you want to last - your server will be <span style="font-weight:bold">deleted after 10 minutes of inactivity</span>. Save often and rember download notebook when you step away (you can always re-upload and start again)!</p>
</div>


## Have fun!

Regardless of outcome, getting to know you is important. Give it your best shot and we'll look forward to following up!

# Exam Questions

## 1. Algo + Data Structures

### Q 1.1: Fibionacci
![fib image](https://upload.wikimedia.org/wikipedia/commons/thumb/9/93/Fibonacci_spiral_34.svg/200px-Fibonacci_spiral_34.svg.png)

#### Q 1.1.1
Given $n$ where $n \in \mathbb{N}$ (i.e., $n$ is an integer and $n > 0$), write a function `fibonacci(n)` that computes the Fibonacci number $F_n$, where $F_n$ is defined by the recurrence relation:

$$ F_n = F_{n-1} + F_{n-2}$$

with initial conditions of:

$$ F_1 = 1,  F_2 = 1$$

In [1]:
def fibonacci(n):
    '''
    1. Recursion
    I will first solve this question in a recursion way. However, for this solution, both fibonacci(n-1) and fibonacci(n-2) need to compute its fibonacci number all the way. 
    That means to get fibonacci(n-1), it will need to compute fibonacci(n-2) and fibonacci(n-3). However, to get fibonacci(n), it will compute fibonacci(n-2) again.
    It is time consuming. Some same computation steps are repeative.
    '''
    assert isinstance(n, int) and n > 0 # n is an integer and n > 0
    
    if n == 1 or n == 2:
        return 1
    return fibonacci(n-1) + fibonacci(n-2) # recursively getting fibonacci(n-1) and fibonacci(n-2)

#### Q 1.1.2
What's the complexity of your implementation?

1. Time Complexity

Let T represent the time complexity,

$T(n) = T(n-1) + T(n-2) < T(n-1) + T(n-1) = 2*T(n-1) = 2*2*T(n-2) = ... = O(2^n)$

2. Space Complexity

This recursion method can be written into a recursion tree. We know in order to get fibonacci(n), we need to extract fibonacci(n-1), fibonacci(n-2),..., and so on. Then, the length of this path is the depth of the tree, which is proportional to n. Each node requests O(1) to store the result. Therefore, the full path requests O(n).

#### Q 1.1.3
Consider an alternative implementation to compute Fibonacci number $F_n$ and write a new function, `fibonacci2(n)`.

In [2]:
def fibonacci2(n):
    '''
    2. Loop + Two Pointers
    For this solution, I will save the fibonacci(n-1) and fibonacci(n-2) as current and previous so that I decrease the time complexity to be O(n).
    The current and previou flags are updated at the same step.
    '''
    assert isinstance(n, int) and n > 0 # n is an integer and n > 0
    
    current = previous = 1
    while n-2 > 0:
        # update the current and previous flag at the same time.
        current, previous = current+previous, current
        n -= 1
    return current

#### Q 1.1.4
What's the complexity of your implementation?

1. Time Complexity

The loop takes n-2 times, so the time complexity is O(n).

2. Space Complexity

These iterative method only requires memory space for the current and previous flag, which will be O(1).

#### Q 1.1.5
What are some examples of optimizations that could improve computational performance?

**Dynamic programming** is a very good but sometimes hard to think of method to deal with this kind of problem. Basically, once we notice that fibonacci(n) is computed repeatedly , it's time to divide the whole problem and use looping to compute each stage. In order not ot repeatedly computing one thing, we will save the current answer for the next step.

**Matrix** is another way and is used quite often in Machine Learning. By converting multiple vectors into a matrix, we can do computation in parallel.

### Q 1.2: Linked List
![ll img](https://upload.wikimedia.org/wikipedia/commons/thumb/6/6d/Singly-linked-list.svg/500px-Singly-linked-list.svg.png)

#### Q 1.2.1
Consider a [singly linked list](https://en.wikipedia.org/wiki/Linked_list), $L$. Write a function `is_palindrome(L)` that detects if $L$ is a [palindrome](https://en.wikipedia.org/wiki/Palindrome), by returning a bool, `True` or `False`.


In [3]:
class Nodes:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = None
        
def is_palindrome(L):
    '''
    1. I will traverse through the whole linked list and save all elements in a new list. Then, I will compare the new list and the reversed new list and see if they are the same.
    
    '''
    new_list = []
    head = L
    while head is not None: # Iteratively extracting each element in the linked list and save them in new_list
        new_list.append(head.val)
        head = head.next
    return new_list == new_list[::-1] # compare new_list with reversed new_list

In [4]:
# example
new = Nodes(1)
new.next = Nodes(2)
new.next.next = Nodes(3)
new.next.next.next = Nodes(2)
new.next.next.next.next = Nodes(1)
# new.next.next.next.next.next = Nodes(1)

is_palindrome(new)

True

#### Q 1.2.2
What is the complexity of your implementation?

1. Time Complexity

It requires O(n) for both traversing through the linked list and comparing the new list.

2. Space Complexity

O(n) for new_list.

#### Q 1.2.3
Consider an alternative implementation to detect if L is a palindrome and write a new function, `is_palindrome2(L)`.

In [5]:
def is_palindrome2(L):
    '''
    2. For this solution, I would like to decrease the space complexity. To do so, I will set two pointers, slow and fast.
       The fast pointer goes two steps ahead each time and the slow pointer goes only one step. That means when the fast one 
       reaches the end, the slow one reaches the mid node. Then, I will start to reverse the linked list from the position of the slow pointer.
       After reversing the linked list, I can now check is_palindrome from the head and the end.
    '''
    slow = fast = L
    while fast is not None and fast.next is not None: # I can find the mid point from the slow pointer when the fast pointer reaches the end.
        slow = slow.next
        fast = fast.next.next
    # starting to reverse the half of the linked list from the position of the slow pointer
    slow = reverse(slow)
    fast = L
    
    # use slow and fast to check is_palindrome respectively from the end and the beginning
    while slow is not None:
        if slow.val != fast.val: # if the elements at slow and fast are different, return False
            return False
        slow = slow.next # if the elements are the same, try to compare the next elements.
        fast = fast.next
    return True
        
def reverse(node):
    prev = None
    while node is not None:
        temp = node.next
        node.next = prev
        prev = node
        node = temp
    return prev

In [6]:
is_palindrome2(new)

True

#### Q 1.2.4
What's the complexity of this implementation?

1. Time Complexity

This function still needs O(n) for traversing the linked list and making comparison.

2. Space Complexity

There is no extra memory needed for the new list in this method. Although I will still use 2 memory space for slow and fast pointers, the space complexity is now O(1).

#### Q 1.2.5 
What are some examples of optimizations that could improve computational performance?


## 2. Prob + Stats

### Q 2.1: Finding $\pi$ in a random uniform?
<img src=https://www.epicurus.com/food/recipes/wp-content/uploads/2015/03/Pi-Day.jpg width="480">

Given a uniform random generator $[0,1)$ (e.g., use your language's standard libary to generate random value), write a a function `compute_pi` to compute [$\pi$](https://en.wikipedia.org/wiki/Pi).

Ans:

This is a statistical simulation problem. Basically, I have a circle inside a square. What I want to compute is that if I drop n random points in the square, what percentage of them will be in the circle area. Notes: n is a large number.

Assuming $r(radius)=1$, the circular area is $\pi*(r^2)=pi$ and the square area is $2*2=4$.

The ratio of circular area and the square area is $\pi/4$.

In [7]:
import numpy as np

In [8]:
def compute_pi(n):
    '''
    My goal is to compute the probability or ratio by droping random data points.
    ''' 
    in_circle = 0
    for _ in range(n):
        x = np.random.uniform()
        y = np.random.uniform()
        if x**2 + y**2 <= 1:
            in_circle += 1
    pi = (in_circle/n)*4
    return pi

In [9]:
print(compute_pi(10000))

3.1516


### Q 2.2: Making a 6-side die roll a 7?

Using a single 6-side die, how can you generate a random number between 1 - 7?

Ans:

First of all, I need to create a distribution in which each of the number among 1 to 7 has the same probability. If I roll the dice twice, they can generate 36 configurations. Then, I can try to make 1 to 7 represent 5 configurations. my suggestion is in the following.

- When the dice is rolled to 1 at the first time and 1 as well at the second, I am required to roll the dice again twice. This represents the rest of 1 configuration which has the probability of $1/36$.

- When the dice is rolled to 1 at the first time and 2\~6 at the second, it is 7 and has the probability of $1/6 * 5/6 = 5/36$.

- When the dice is rolled to 2\~6 at the first time and any number at the second, I will let the second number represent the final answer. This probability of $5/6 * 1/6 = 5/36$ is for 1\~6.

Then, all of the number is generated uniformly and randomly with a probability of $5/36$.

### Q 2.3: Is normality uniform?

<img src=https://rednaxela1618.files.wordpress.com/2014/06/uniformnormal.png width="480">


Given draws from a normal distribution with known parameters, how can you simulate draws from a uniform distribution?

### Q 2.4: Should you pay or should you go?

![coin flip](https://lh5.ggpht.com/iwD6MnHeHVAXNBgrO7r4N9MQxxYi6wT9vb0Mqu905zTnNlBciONAA98BqafyjzC06Q=w300)

Let’s say we play a game where I keep flipping a coin until I get heads. If the first time I get heads is on the nth coin, then I pay you $2^{(n-1)}$ US dollars. How much would you pay me to play this game? Explain.

### Q 2.5: Uber vs. Lyft

![uber vs lyft](http://usiaffinity.typepad.com/.a/6a01347fc1cb08970c01bb0876bcbe970d-pi)

You request 2 UberX’s and 3 Lyfts. If the time that each takes to reach you is IID, what is the probability that all the Lyfts arrive first? What is the probability that all the UberX’s arrive first?

### Q 2.6: Pick your prize
<img src=https://miro.medium.com/max/1100/1*m5b3O9sE68UCXjLw5oxy2g.png width="480">

A prize is placed at random behind one of three doors and you are asked to pick a door. To be concrete, say you always pick door 1. Now the game host chooses one of door 2 or 3, opens it and shows you that it is empty. They then give you the option to keep your picked door or switch to the unopened door. Should you stay or switch if you want to maximize your probability of winning the prize?

## 3 Conceptual ML

### Q 3.1 Why study gradient boosting or neural networks?

Consider a regression setting where $X \in \mathbb{R}^p$ and $Y \in \mathbb{R}$. The goal is to come up with a function $f(X): \mathbb{R}^p \rightarrow \mathbb{R}$ that minimizes the squared-error loss $(Y - f(X))^2$. Since X, Y are random variables, we seek to minimize the expectation of the squared error loss as follows
\begin{equation}
EPE(f) = \mathbb{E}\left[(Y-f(X)^2\right]
\end{equation}
where EPE stands for expected prediction error. One can show that minimizing the expected prediction error leads to the following _regression function_
\begin{equation}
f(x) = \mathbb{E}\left[Y|X=x\right]
\end{equation}

The goal of any method is to approximate the regression function above, which we denote as $\hat{f}(x)$. For example, linear regression explicitly assumes that the regression function is approximately linear in its arguments, i.e. $\hat{f}(x) = x^T\beta$ while a neural network provides a nonlinear approximation of the regression function. 

The simplest of all these methods is [k-nearest neighbors](https://en.wikipedia.org/wiki/K-nearest_neighbors_algorithm). Given $x$ and some neighbourhood of $k$ points $N_k(x)$, $\hat{f}(x)$ is simply the average of all $y_i|x_i \in N_k(x)$.  Let $N$ denote the training sample size. Under mild regularity conditions on the joint probability distribution $Pr(X, Y)$, one can show that as $N \rightarrow \infty$, $k \rightarrow \infty$ such that $k/N \rightarrow 0$, then $\hat{f}(x) \rightarrow f(x)$ where $\rightarrow$ means approaches or goes to. In other words, the k-nearest neighbors algorithm converges to the ideal solution as both the training sample size and number of neighbors increase to infinity.

Now given this _universal approximator_, why look any further and research other methods? Please share your thoughts.


### Q 3.2 Model Selection and Assesment

Consider a multiclass classification problem with a large number of features $p >> N$, for e.g $p=10000, N=100$ The task is threefold
1. Find a "good" subset of features that show strong _univariate_ correlation with class labels
2. Using the "good" subset, build a multi class classifier
3. Estimate the generalization error of the final model

Given this dataset, outline your approach and please be sure to cover the following
- Data splitting
- Model Selection: either estimating the performance of different classifiers or the same classifier with different hyperparameters
- Model Assessment: having chosen a classifier, estimating the generalization error

Assume all features are numerical, the dataset contains no NULLS, outliers, etc. and doesn't require any preprocessing.



Ans:

1. Data Splitting

Before splitting the full dataset, I will first take a look at the distribution of the target variable in case that there exists the imbalanced dataset problem. Considering the size of the dataset is pretty small, first of all, I will randomly split the full dataset into two subsets, training and testing sets, with a percentage of 80% and 20%. In the training set, I will do 10-fold cross validation so that it will further be randomly splitted into 10 training-validation subsets. At this step, some upsampling or undersampling methods may be applied on the training set if the imbalanced problem exists.

2. Dimension Reduction

Based on the assumption of this dataset, the feature size is way bigger than the sample size. Curse of dimenionality, which not only leads to high computation cost but also over-fitting problems, would happen here. There are some techniques I can try.

 - No feature selection: This could cause overfitting.
 - Filter method: Some features with high linear correlation to each other are redundant. A part of them can be removed. Also, this specific method also measures linear relationship.
 - Greedy: Backward and forward feature selection methods will take a high computation cost. I usually don't use them.
 - Regularization and tree-based models: These methods are directly connected with the models. For example, we can use decision trees on our numerical features to get feature importance. Besides, l1 and l2 regularization with logistic regression, SVM, and multilayer perceptron are common methods to reduce overfitting.
 - Dimension reduction techniques: Methods, like PCA, Multidimensional scaling, generate a new subset of features from the original feature sets.
 
The methods above should be deeply explored in this hign-dimension dataset.

3. Model Selection

I would set up the pipeline starting from dimension reduction and model training. For model training, grid search and 10-fold cross validation for tuning hyperparameters are applied here. I will first try a wide range of hyperparameters based on grid search and then focus on a limited range by observing the metrics from 10-fold cross validation. Multiple models will be tested to see which one works better in this dataset.

4. Model Assessment

Since this is a classification problem, the metrics for regression models, such as MSE and MAE, are not suitable. Instead, accuracy, ROC, and f1-score should be applied. However, accuracy can be misleading according to imbalanced class distribution. Therefore, I would prefer to use ROC as the metrics when implementing cross validation.

Model reliability and generalization ability are important during model assessment, I will observe learning rate for training loss and validation loss when training the model to avoid overfitting or underfitting.